## Phase 1: ML Foundation (Day 1–20)
# DAY 6 — Data Cleaning 

## Data Cleaning (SUPER IMPORTANT)
### Concept (Real World Truth)
#### 👉 Real data is never clean

**You will face:**

* Missing values
* Wrong values
* Empty cells



In [1]:
# Example Dataset
data = {
    "Hours": [1, 2, None, 4, 5],
    "Practice": [10, None, 30, 40, 50],
    "Marks": [15, 30, 45, None, 75]
}

## Task : 1
### Answer this:

### 1. What problems can missing values cause in model training?
### 2. What can we do with missing values? (2 ways)

### 1. Missing values can reduce accuracy and cause errors in model training.

### 2. Two ways to handle:
* - Remove missing values (dropna)
* - Fill missing values (fillna using mean/median/mode)

## Coding Task

In [3]:
# Use this dataset:

data = {
    "Hours": [1, 2, None, 4, 5],
    "Practice": [10, None, 30, 40, 50],
    "Marks": [15, 30, 45, None, 75]
}

## Task : 2

### 🔹 Step 1: Create DataFrame

### 🔹 Step 2: Check missing values
df.isnull().sum()

### 🔹 Step 3: Handle missing values
👉 Do BOTH:

#### Method 1: Drop

df_drop = df.dropna()
#### Method 2: Fill

df_fill = df.fillna(df.mean())

In [5]:
# DataFrame
import pandas as pd

df = pd.DataFrame(data)
df

,Hours,Practice,Marks
0,1.0,10.0,15.0
1,2.0,NaN,30.0
2,NaN,30.0,45.0
3,4.0,40.0,NaN
4,5.0,50.0,75.0


In [6]:
# Checking missing Values

df.isnull().sum()

Hours       1
Practice    1
Marks       1
dtype: int64

In [7]:
# Handling Missing Values

In [8]:
# Method : 1 Drop

df.dropna()

,Hours,Practice,Marks
0,1.0,10.0,15.0
4,5.0,50.0,75.0


In [10]:
# Method : 2 Fill

df.fillna(df.mean())

,Hours,Practice,Marks
0,1.0,10.0,15.00
1,2.0,32.5,30.00
2,3.0,30.0,45.00
3,4.0,40.0,41.25
4,5.0,50.0,75.00


## Mini Task
### 👉 Compare:

* Shape before cleaning
* After drop
* After fill



In [12]:
print("Missing Values : ",df.isnull().sum())

Missing Values :  Hours       1
Practice    1
Marks       1
dtype: int64


In [13]:
print("Shape (original) : ",df.shape)

Shape (original) :  (5, 3)


In [15]:
print("Shape (dropna) : ",df.dropna().shape)

Shape (dropna) :  (2, 3)


In [17]:
print("Shape (fillna) : ",df.dropna().shape)

Shape (fillna) :  (2, 3)


## Which is better in most real-world cases:
### dropna OR fillna ?
# fillna is better because it keeps all data and avoids data loss, which is important for model training.

## FINAL Mini Project
#### Now we combine everything 👇
### Your Task (Real ML Flow)

In [18]:
# Use the same dirty dataset:

data = {
    "Hours": [1, 2, None, 4, 5],
    "Practice": [10, None, 30, 40, 50],
    "Marks": [15, 30, 45, None, 75]
}

In [ ]:
# 1. Create DataFrame
# 2. Handle missing values
# 3. Define features
# 4. Train model
# 5. Predict

In [19]:
# 1. Create DataFrame

df = pd.DataFrame(data)
df

,Hours,Practice,Marks
0,1.0,10.0,15.0
1,2.0,NaN,30.0
2,NaN,30.0,45.0
3,4.0,40.0,NaN
4,5.0,50.0,75.0


In [21]:
# 2. Handle missing values

df = df.fillna(df.mean())

In [22]:
# 3. Define features

X = df[['Hours', 'Practice']]
y = df['Marks']

In [27]:
# 4. Train model

from sklearn.linear_model import LinearRegression
model = LinearRegression()
model.fit(X, y)

LinearRegression()

In [28]:
# 5. Predict

new_data = pd.DataFrame({
    "Hours": [6],
    "Practice": [60]
})

print(model.predict(new_data))

[80.89285714]


## 👉 Why did we clean data before training, not after?

### We clean data before training because missing values can cause errors and reduce model accuracy. 
### Clean data helps the model learn correct patterns.

##  Important Insight (Don’t Miss This)

### Earlier (clean dataset):
* Prediction = 90

### Now (after fillna):
* Prediction ≈ 80.89
* 
### 👉 Why difference?

✔️ Because:

* Missing values were replaced with mean
* Data pattern changed slightly
* Model learned a different relationship

### 👉 This is real-world ML behavior

